mlflow/types/utils.py

```diff
@@ -79,7 +79,75 @@ def clean_tensor_type(dtype: np.dtype):
     return dtype
 
 
-def _infer_datatype(data) -> DataType:
+def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
+    """
+    Infer an MLflow Colspec type from the dataset.
+
+    :param data: data to infer from.
+    :return: Object
+    """
+    dtype = _infer_datatype(data)
+    _validate_inferred_type(dtype)
+    return dtype
+
+
+def _infer_datatype(data: Any) -> Union[DataType, Array, Object]:
+    if isinstance(data, dict):
+        properties = [Property(name=k, dtype=_infer_datatype(v)) for k, v in data.items()]
```

Comment:
What if we have `{"a": []}`, then this would raise exception at `Property` construction stage.

Reply:
Right, I think that's what we want. Property's value has to be non-empty so we can infer type of it.

Reply:
I think we should raise a clearer message instead of a construction error. Say if `_infer_datatype(v)` is None, we directly raise MlflowException here

Reply:
That's true, will add explicit check here to raise more friendly message.